components/webext-storage/src/migration.rs

```diff
@@ -0,0 +1,335 @@
+                return None;
+            }
+            _ => {
+                log::trace!("key doesn't exist in the map");
+                return None;
+            }
+        };
+        let data = match record_map.remove("data") {
+            Some(d) => d,
+            _ => {
+                log::trace!("data doesn't exist in the map");
+                return None;
+            }
+        };
+        Some(Parsed { ext_id, key, data })
+    }
+}
+
+// The info we parse from the raw DB strings.
+struct Parsed<'a> {
+    ext_id: &'a str,
+    key: String,
+    data: serde_json::Value,
+}
+
+pub fn migrate(tx: &Transaction<'_>, filename: &PathBuf) -> Result<usize> {
+    // We do the grouping manually, collecting string values as we go.
+    let mut last_ext_id = "".to_string();
+    let mut curr_values: Vec<(String, serde_json::Value)> = Vec::new();
+    let mut curr_value_len = 0;
+    let mut num_extensions = 0;
+    for row in read_rows(filename)? {
+        log::trace!("processing '{}' - '{}'", row.col_name, row.record);
+        let parsed = match row.parse() {
+            Some(p) => p,
+            None => continue,
+        };
+        // Do our "grouping"
+        if parsed.ext_id != last_ext_id {
+            if last_ext_id != "" && !curr_values.is_empty() {
+                // a different extension id - write what we have to the DB.
+                if do_insert(tx, &last_ext_id, curr_values) {
+                    num_extensions += 1;
+                }
+            }
+            last_ext_id = parsed.ext_id.to_string();
+            curr_values = Vec::new();
+            curr_value_len = 0;
+        }
+        // no 'else' here - must also enter this block on ext_id change.
+        if parsed.ext_id == last_ext_id {
+            // Do our "quota but not a quota" thing.
+            let new_len = curr_value_len + parsed.data.to_string().len() + parsed.key.len();
+            if new_len < MAX_LEN && curr_values.len() < MAX_KEYS {
+                curr_values.push((parsed.key.to_string(), parsed.data));
+                curr_value_len = new_len;
+            }
+            log::trace!(
+                "extension {} now has {} bytes, {} keys",
+                parsed.ext_id,
+                curr_value_len,
+                curr_values.len()
+            );
+        }
+    }
+    // and the last one
+    if last_ext_id != "" && !curr_values.is_empty() {
+        // a different extension id - write what we have to the DB.
+        if do_insert(tx, &last_ext_id, curr_values) {
+            num_extensions += 1;
+        }
+    }
+    Ok(num_extensions)
+}
+
+fn read_rows(filename: &PathBuf) -> Result<Vec<LegacyRow>> {
+    let flags = OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_READ_ONLY;
+    let src_conn = Connection::open_with_flags(&filename, flags)?;
+    let mut stmt = src_conn.prepare(
+        "SELECT collection_name, record FROM collection_data
+         ORDER BY collection_name",
+    )?;
+    let rows = stmt
```

Comment:
We'll need initial and final counts of these for telemetry... Maybe we can just get them later though. But probably not counting non-default/ would be good.

But I guess that can be done later, too.